File: php/test/AiJobEntityTest.php
<?php
declare(strict_types=1);

// AiJob entity test

require_once __DIR__ . '/../memesiocontentcreation_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class AiJobEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = MemesioContentCreationSDK::test(null, null);
        $ent = $testsdk->AiJob(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = ai_job_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "ai_job." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set MEMESIO_CONTENT_CREATION_TEST_AI_JOB_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $ai_job_ref01_ent = $client->AiJob(null);
        $ai_job_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.ai_job"), "ai_job_ref01"));

        $ai_job_ref01_data_result = $ai_job_ref01_ent->create($ai_job_ref01_data, null);
        $ai_job_ref01_data = Helpers::to_map(is_object($ai_job_ref01_data_result) && method_exists($ai_job_ref01_data_result, 'data_get') ? $ai_job_ref01_data_result->data_get() : $ai_job_ref01_data_result);
        $this->assertNotNull($ai_job_ref01_data);
        $this->assertNotNull($ai_job_ref01_data["id"]);

        // LOAD
        $ai_job_ref01_match_dt0 = [
            "id" => $ai_job_ref01_data["id"],
        ];
        $ai_job_ref01_data_dt0_loaded = $ai_job_ref01_ent->load($ai_job_ref01_match_dt0, null);
        $ai_job_ref01_data_dt0_load_result = Helpers::to_map(is_object($ai_job_ref01_data_dt0_loaded) && method_exists($ai_job_ref01_data_dt0_loaded, 'data_get') ? $ai_job_ref01_data_dt0_loaded->data_get() : $ai_job_ref01_data_dt0_loaded);
        $this->assertNotNull($ai_job_ref01_data_dt0_load_result);
        $this->assertEquals($ai_job_ref01_data_dt0_load_result["id"], $ai_job_ref01_data["id"]);

    }
}

function ai_job_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/ai_job/AiJobTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = MemesioContentCreationSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["ai_job01", "ai_job02", "ai_job03", "job01", "job02", "job03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("MEMESIO_CONTENT_CREATION_TEST_AI_JOB_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "MEMESIO_CONTENT_CREATION_TEST_AI_JOB_ENTID" => $idmap,
        "MEMESIO_CONTENT_CREATION_TEST_LIVE" => "FALSE",
        "MEMESIO_CONTENT_CREATION_TEST_EXPLAIN" => "FALSE",
        "MEMESIO_CONTENT_CREATION_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["MEMESIO_CONTENT_CREATION_TEST_AI_JOB_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["MEMESIO_CONTENT_CREATION_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["MEMESIO_CONTENT_CREATION_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new MemesioContentCreationSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["MEMESIO_CONTENT_CREATION_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["MEMESIO_CONTENT_CREATION_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: lua/test/ai_job_entity_test.lua
-- AiJob entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("memesio-content-creation_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("AiJobEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:AiJob(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = ai_job_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "ai_job." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set MEMESIO_CONTENT_CREATION_TEST_AI_JOB_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local ai_job_ref01_ent = client:AiJob(nil)
    local ai_job_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.ai_job"), "ai_job_ref01"))

    local ai_job_ref01_data_result, err = ai_job_ref01_ent:create(ai_job_ref01_data, nil)
    assert.is_nil(err)
    ai_job_ref01_data = helpers.to_map(type(ai_job_ref01_data_result) == 'table' and ai_job_ref01_data_result.data_get and ai_job_ref01_data_result:data_get() or ai_job_ref01_data_result)
    assert.is_not_nil(ai_job_ref01_data)
    assert.is_not_nil(ai_job_ref01_data["id"])

    -- LOAD
    local ai_job_ref01_match_dt0 = {
      id = ai_job_ref01_data["id"],
    }
    local ai_job_ref01_data_dt0_loaded, err = ai_job_ref01_ent:load(ai_job_ref01_match_dt0, nil)
    assert.is_nil(err)
    local ai_job_ref01_data_dt0_load_result = helpers.to_map(type(ai_job_ref01_data_dt0_loaded) == 'table' and ai_job_ref01_data_dt0_loaded.data_get and ai_job_ref01_data_dt0_loaded:data_get() or ai_job_ref01_data_dt0_loaded)
    assert.is_not_nil(ai_job_ref01_data_dt0_load_result)
    assert.are.equal(ai_job_ref01_data_dt0_load_result["id"], ai_job_ref01_data["id"])

  end)
end)

function ai_job_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/ai_job/AiJobTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read ai_job test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "ai_job01", "ai_job02", "ai_job03", "job01", "job02", "job03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("MEMESIO_CONTENT_CREATION_TEST_AI_JOB_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["MEMESIO_CONTENT_CREATION_TEST_AI_JOB_ENTID"] = idmap,
    ["MEMESIO_CONTENT_CREATION_TEST_LIVE"] = "FALSE",
    ["MEMESIO_CONTENT_CREATION_TEST_EXPLAIN"] = "FALSE",
    ["MEMESIO_CONTENT_CREATION_APIKEY"] = "",
  })

  local idmap_resolved = helpers.to_map(
    env["MEMESIO_CONTENT_CREATION_TEST_AI_JOB_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["MEMESIO_CONTENT_CREATION_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      -- FIRST, so the generated fields below win: sdk-test-control.json's
      -- test.client.options adds to the live client, it does not redirect it.
      runner.live_client_options(),
      {
        apikey = env["MEMESIO_CONTENT_CREATION_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["MEMESIO_CONTENT_CREATION_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["MEMESIO_CONTENT_CREATION_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
